Jenkinsfile

```diff
@@ -1,175 +1,60 @@
-*/
-def runTox(environment) {
-  def processes = env.PYTEST_PROCESSES ?: 'auto'
-  try {
-    wrap([$class: 'AnsiColorBuildWrapper']) {
-      withCredentials([[
-        $class: 'StringBinding',
-        credentialsId: 'SAUCELABS_API_KEY',
-        variable: 'SAUCELABS_API_KEY']]) {
-        withEnv(["PYTEST_ADDOPTS=${PYTEST_ADDOPTS} " +
-          "-n=${processes} " +
-          "--base-url=${PYTEST_BASE_URL} " +
-          "--driver=SauceLabs " +
-          "--variables=capabilities.json " +
-          "--color=yes"]) {
-          sh "tox -e ${environment}"
-        }
+      try {
+        config = readYaml file: "jenkins.yml"
       }
-    }
-  } catch(err) {
-    currentBuild.result = 'FAILURE'
-    throw err
-  } finally {
-    dir('results') {
-      stash environment
-    }
-  }
-}
-
-/** Send a notice to #fxtest-alerts on irc.mozilla.org with the build result
- *
- * @param result outcome of build
-*/
-def ircNotification(result) {
-  def nick = "fxtest${BUILD_NUMBER}"
-  def channel = '#fx-test-alerts'
-  result = result.toUpperCase()
-  def message = "Project ${JOB_NAME} build #${BUILD_NUMBER}: ${result}: ${BUILD_URL}"
-  node {
-    sh """
-        (
-        echo NICK ${nick}
-        echo USER ${nick} 8 * : ${nick}
-        sleep 5
-        echo "JOIN ${channel}"
-        echo "NOTICE ${channel} :${message}"
-        echo QUIT
-        ) | openssl s_client -connect irc.mozilla.org:6697
-    """
-  }
-}
+      catch (e) {
+        config = []
+      }
+      println "config ==> ${config}"
 
-stage('Checkout') {
-  node {
-    timestamps {
-      deleteDir()
-      checkout scm
-      stash 'workspace'
+      if (!config || (config && config.pipeline && config.pipeline.enabled == false)) {
+        println "Pipeline disabled."
+      }
     }
-  }
-}
 
-def builders = [:]
-for (entry in entrySet(environments)) {
-  def environment = entry.key
-  def capabilities = entry.value
-  builders[(environment)] = {
-    node {
-      timeout(time: 1, unit: 'HOURS') {
-        timestamps {
-          deleteDir()
-          unstash 'workspace'
-          writeCapabilities(capabilities)
-          withCredentials([[
-            $class: 'FileBinding',
-            credentialsId: 'SUMO_VARIABLES',
-            variable: 'VARIABLES']]) {
-            withEnv(["PYTEST_ADDOPTS=--variables=${VARIABLES}"]) {
-              runTox(environment)
-            }
-          }
-        }
+    docker_image = "${config.project.docker_name}:${GIT_COMMIT_SHORT}"
+
+    stage("Build") {
+      if (!dockerImageExists(docker_image)) {
+            sh "echo 'ENV GIT_SHA ${GIT_COMMIT}' >> Dockerfile"
+
+            // Docker always copies data as root. Chowning files to root before
+            // the build command will prevent docker cache invalidation due to
+            // different file metadata. When we upgrade to a newer docker
+            // version we should probably use --chown flag for COPY. See
+            // https://github.com/moby/moby/pull/34263
```

Comment:
This is fascinating. Thanks! I've been wondering why this didn't seem to work quite right sometimes. But it does seem to use cache for me locally. Why is that?

Reply:
Ah! I get it. It works fine from the same system, but when building from another system it'll not use cache because it was built with a different user on a different machine. Interesting.

Reply:
Yeah it's... complicated. I had this question too and  it cost me a fair amount to time to get this working.